src/v/kafka/server/handlers/describe_producers.cc

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "kafka/server/handlers/describe_producers.h"
+
+#include "cluster/errc.h"
+#include "cluster/partition_manager.h"
+#include "cluster/shard_table.h"
+#include "kafka/protocol/describe_producers.h"
+#include "kafka/protocol/errors.h"
+#include "kafka/protocol/schemata/describe_producers_response.h"
+#include "kafka/server/handlers/details/security.h"
+#include "kafka/server/request_context.h"
+#include "kafka/server/response.h"
+#include "model/fundamental.h"
+#include "model/namespace.h"
+#include "resource_mgmt/io_priority.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/when_all.hh>
+
+namespace kafka {
+
+namespace {
+
+int64_t get_time_since_epoch(ss::lowres_clock::time_point last_update_time) {
+    auto diff = ss::lowres_clock::now() - last_update_time;
+    return std::chrono::system_clock::now().time_since_epoch().count()
+           - diff.count();
+}
+
+ss::future<checked<cluster::rm_stm::transaction_set, kafka::error_code>>
+get_producers_for_partition(request_context& ctx, model::ntp ntp) {
+    auto shard = ctx.shards().shard_for(ntp);
+    if (!shard) {
+        vlog(
+          klog.debug,
+          "Can not find shard for ntp({}) to get info about producers",
+          ntp);
+        co_return kafka::error_code::not_leader_for_partition;
+    }
+
+    co_return co_await ctx.partition_manager().invoke_on(
+      *shard,
+      [ntp](cluster::partition_manager& pm)
+        -> ss::future<
+          checked<cluster::rm_stm::transaction_set, kafka::error_code>> {
+          auto partition = pm.get(ntp);
+          if (!partition) {
+              co_return kafka::error_code::unknown_topic_or_partition;
+          }
```

Comment:
shortly after this PR was posted we added static analysis checks into CI that prevent coroutine lambdas from being introduced. that is why one the CI jobs for this PR is failing. you'll need to either factor this out into a separate function (i think this would be preferred), or rewrite it without coroutines (usually makes sense when the lambda is tiny).